collect_app/src/main/java/org/odk/collect/android/javarosawrapper/JavaRosaFormController.java

```diff
@@ -396,15 +396,19 @@ public int answerQuestion(FormIndex index, IAnswerData data) throws JavaRosaExce
     }
 
     public ValidationResult validateAnswers(boolean markCompleted) throws JavaRosaException {
-        ValidateOutcome validateOutcome = getFormDef().validate(markCompleted);
-        if (validateOutcome != null) {
-            this.jumpToIndex(validateOutcome.failedPrompt);
-            if (indexIsInFieldList()) {
-                stepToPreviousScreenEvent();
+        try {
+            ValidateOutcome validateOutcome = getFormDef().validate(markCompleted);
+            if (validateOutcome != null) {
+                this.jumpToIndex(validateOutcome.failedPrompt);
+                if (indexIsInFieldList()) {
+                    stepToPreviousScreenEvent();
+                }
+                return new FailedValidationResult(validateOutcome.failedPrompt, validateOutcome.outcome);
             }
-            return new FailedValidationResult(validateOutcome.failedPrompt, validateOutcome.outcome);
+            return SuccessValidationResult.INSTANCE;
+        } catch (RuntimeException e) {
```

Comment:
I did here exactly what we already do in `stepToPreviousScreenEvent`/`stepToNextScreenEvent` just catch `RuntimeException` and turn it into `JavaRosaException` so that then it's handled properly and an error dialog is displayed.